File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==NULL || k==0) return head;
        ListNode* last = head;
        int n = 1;
        while(last->next!=NULL) {
            n++;
            last = last->next;
        }

        k = k%n;
        if(k==n || k==0) return head;

        ListNode *breakNode = head;
        int times = n-k;
        while(--times) {
            breakNode = breakNode->next;
        }
        ListNode* start = breakNode->next;

        last->next = head;
        breakNode->next = NULL;
        head = start;
        return head;
    }
};

File: 61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """
        if head == None:
            return head

        last = head
        n = 1
        while(last.next!=None):
            n+=1
            last = last.next
        
        k = k%n
        if k==0 or k==n:
            return head
        

        breakNode = head
        for i in range(1, (n-k)):
            breakNode = breakNode.next
        
        start = breakNode.next

        last.next = head
        breakNode.next = None
        head = start
        return head

        